### ISSUE TITLE
Callback with add_loss method in Keras

### ISSUE THREAD

**Issue by RCTimms** [keras-team/keras#13826]
**System information**  
- Have I written custom code (as opposed to using example directory):  
Yes
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04):  
Google Colab
- TensorFlow backend (yes / no):  
Yes
- TensorFlow version:  
2.1.0
- Keras version:  
2.2.4
- Python version:  
3.6.9

**Describe the current behavior**  
I am trying to implement a VAE style network in Keras. I compute my negative log-likelihood and KL divergence and add them to my model with model.add_loss(NLL) and model.add_loss(KL), respectively.

I would like to scale my KL term as training progresses ("KL-annealing"). I attempted to do this with custom callback ([detailed here](https://deepakbaby.github.io/post/vae-insights/)), but the KL loss term does not get updated - the model.add_loss(KL) term is constant over time, despite the KL weight getting updated (see Figure).

**Is this a limitation of the model.add_loss() functionality? Can it not interact with variables which vary over training epochs?**

**Describe the expected behavior**  
KL contribution scales with the KL weight

**Code to reproduce the issue**  
```
...
<NLL calculations>
...

# Add the first loss to the model, the NLL:
model.add_loss(NLL)

from keras.callbacks import Callback

klstart = 40
# number of epochs over which KL scaling is increased from 0 to 1
kl_annealtime = 20

weight = tf.keras.backend.variable(0.0) #intialiase KL weight term
class AnnealingCallback(Callback):
    def __init__(self, weight):
        self.weight = weight
    def on_epoch_end (self, epoch, logs={}):
        if epoch > klstart :
            new_weight = min(tf.keras.backend.get_value(self.weight) + (1./ kl_annealtime), 1.)
            tf.keras.backend.set_value(self.weight, new_weight)
        print ("Current KL Weight is " + str(tf.keras.backend.get_value(self.weight)))

# Now the KL divergence:
<KL weight calculations>
KL = weight*tf.reduce_mean(tfp.distributions.kl_divergence(p, q))
model.add_loss(KL)

# Now compile the model with a specified optimiser
opt = tf.keras.optimizers.Adam(lr=0.001,clipnorm=0.1)
model.compile(optimizer=opt)

# Monitor how the NLL and KL divergence differ over time
model.add_metric(KL, name='kl_loss', aggregation='mean')
model.add_metric(NLL, name='mse_loss', aggregation='mean')

ops.reset_default_graph()
history=model.fit(Y_portioned, # Input or "Y_true"
                    verbose=1,
                    callbacks=[earlystopping_callback,callback_reduce_lr,AnnealingCallback(weight)],
                    epochs=650,
                    batch_size=8
                    ) # <- Increase batch size for speed up
``` 
Many thanks in advance

![image](https://user-images.githubusercontent.com/45631644/75178802-ae261700-5730-11ea-9d94-d288abdd2911.png)

**Comment by ben-arnao** [comment]
The calculation for loss is generally not something that changes with the number of epochs/batches. Are you sure you this is what you want to do?

**Comment by RCTimms** [comment]
Thanks for the response @ben-arnao. In short, yes this is definitely what I want to do - but only by a scaling factor.

KL annealing is a very standard procedure and is well documented (see [here] for example(https://arxiv.org/abs/1511.06349)).

The loss is made up of two components, 
L = NLL + weight*KL

and it is only the contribution of the KL term that I want to change as a function of epoch number. My problem appears to be that my call to `model.add_loss(KL)` is static, such that changing the `weight` variable has no materialistic effect on the KL loss calculation.

Hope that clarifies things for you.

**Comment by ben-arnao** [comment]
It seems like you just want a metric, not a loss. There is only one loss the model is compiled with and the optimizer tries to reduce this, metrics are just extra information you want to see. That being said i don't think add_loss is what you want. You're best bet is to modify your metric directly in your callback so you have access to the # of trials.

**Comment by RCTimms** [comment]
@ben-arnao 
Okay, thanks. Sorry if my language was misleading. I want my metric, `weight` to influence my total loss. 

> That being said i don't think add_loss is what you want. You're best bet is to modify your metric directly in your callback so you have access to the # of trials.

Hmm okay - thanks. Am I not doing this with my callback? The Figure shows that `weight` does get updated over training epochs, it just doesn't propagate through to the `KL` loss. Am I missing something or doing something stupid?

**Comment by ben-arnao** [comment]
> @ben-arnao
> Okay, thanks. Sorry if my language was misleading. I want my metric, `weight` to influence my total loss.
> 
> > That being said i don't think add_loss is what you want. You're best bet is to modify your metric directly in your callback so you have access to the # of trials.
> 
> Hmm okay - thanks. Am I not doing this with my callback? The Figure shows that `weight` does get updated over training epochs, it just doesn't propagate through to the `KL` loss. Am I missing something or doing something stupid?

Do we know what the actual output of 

`weight*tf.reduce_mean(tfp.distributions.kl_divergence(p, q))`

It would seem that is was is actually being reported for your metric 'kl_loss'

**Comment by RCTimms** [comment]
@ben-arnao just updated my previous comment. Sorry if we cross-posted.

> Do we know what the actual output this is?
>weight*tf.reduce_mean(tfp.distributions.kl_divergence(p, q))
>It would seem that is what is actually being reported for your metric 'kl_loss'

Yes. For the data I am using the magnitude of `tf.reduce_mean(tfp.distributions.kl_divergence(p, q))` is greater than 0. 

`weight` gets initialised to be `0.0` and so this is why the metric `kl_loss` starts at 0 as in the attached screenshot.

However, when `weight` becomes >0, the `kl_loss` does not become non-zero. In other words, it updates locally, but does not have any effect on the actual loss calculation.

**Comment by ben-arnao** [comment]
Well your NLL loss is translating to the mse metric so my bet would be the term above is in fact zero since we already know the weight should be nonzero

**Comment by RCTimms** [comment]
@ben-arnao whoops, that's a typo in my code
`model.add_metric(NLL, name='mse_loss', aggregation='mean')`
should be
`model.add_metric(NLL, name='NLL', aggregation='mean')`

I feel like we're getting away from the issue at hand though...

**Comment by bmxitalia** [comment]
@RCTimms I'm implementing a variational autoencoder with KL annealing and I'm having the same issue. It seems that one time that the model is compiled it isn't possible to change its loss during training. I tried defining a custom loss with a function that takes in input the KL term but since the process start with KL term at 0.0 the entire training process is performed with that value. I tried to use a K.variable for the KL term like you but nothing is changed. I don't know if it is possible to use keras at low level like PyTorch. With PyTorch is very easy to implement KL annealing. Did you find a solution for Keras? Thank you!

**Comment by RCTimms** [comment]
Hi @bmxitalia 

No joy as of yet (I have been distracted with other things!), but still something I'd like to solve. Will let you know if I crack it.

Best of luck

**Comment by bmxitalia** [comment]
@RCTimms it seems I found a way to fix the issue. 
You have to declare a K.variable() for your kl_weight (kl_weight = K.variable(0.0)) and then create a custom loss function in the form:
def loss(y_true, y_pred):
     rec_loss = ...
     KL_loss = ...
     return rec_loss + kl_weight*KL_loss

Remember to compile the model with model.compile(x, y, loss=loss).

The only thing you have to keep in mind is that the K.variable() has been declared before the loss function declaration.
Then you should create a custom training loop, like this one:

for i in range(n_epochs):
    for j in range(n_batch):
        //here you should have a batch created for your model, you have 
        //to create a sort of batch generator with this loop
        model.train_on_batch(batch_you_have_created)
        //update your kl_weight with K.set_value() at the value you want
    //now, here you can evaluate your model

For me this scheme has worked. 
If you want a way to debug the losses you can create custom metrics for that, where you simply return the part of the loss you are interested in debugging.
If you need an example I can send you my Collaborative filtering VAE implementation that use this scheme. Hope this can help you!

**Comment by giovp** [comment]
I'm experiencing the same exact problem with TF2.3

**Comment by giovp** [comment]
I picked it up again and wasted way too many hours on it (as usual with TF), but this is my solution:
Here, "self" is just a generic python class.
```python
class KLWeightLayer(tf.keras.layers.Layer):
            def __init__(
                self,
                weight: tf.Variable,
                name="klweight_layer",
                **kwargs,
            ):
                super().__init__(name=name, **kwargs)

                self.aw = weight
            def call(self, inputs):
                out = tf.math.multiply(self.aw, tf.math.reduce_sum(inputs), name="kl_scaled")
                return out

self.encoder_model = tf.keras.Model(
            inputs=self.encoder_input, outputs=qz_x.sample(), name="encoder",
        )
self.aw = tf.Variable(self.annealweight, trainable=False, name="aw", validate_shape=False)
scaled_kl = KLWeightLayer(self.aw)(kl_divergence)#tf.math.multiply(aw, tf.math.reduce_sum(kl_divergence), name="kl_scaled")

self.encoder_model.add_loss(
            scaled_kl
        )
self.encoder_model.add_metric(scaled_kl, name="avg_kl", aggregation="mean")
```
The callback I'm using is just a generic callback subclass
```python
class LinearAnnealingCallback(tf.keras.callbacks.Callback):
    def __init__(
        self,
        weight: tf.Variable,
        final_value: float,
        epochs: int = None,
    ):

        self.weight = weight
        self.final_value = final_value
        self.annealtime = int(round(epochs*0.5))
        self.annealstart = 0

    def on_epoch_end(self, epoch, logs={}):
        if epoch >= self.annealstart:
            new_weight = min(
                tf.keras.backend.get_value(self.weight) + (self.final_value / self.annealtime),
                self.final_value,
            )
            tf.keras.backend.set_value(self.weight,new_weight)
        print("\n Current Anneal Weight is " + str(tf.keras.backend.get_value(self.weight)))
```

I think it has to do with the fact that if tf.Variable is technically not in the graph, unless you assign it to an actual layer of the model (I'm using keras functional API). HTH

**Comment by sushreebarsa** [comment]
@RCTimms  Moving this issue to closed status as there has been no recent activity, in case you still face the error please create a new issue.Thanks!
